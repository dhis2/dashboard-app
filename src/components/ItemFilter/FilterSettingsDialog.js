import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { DIMENSION_ID_PERIOD, DIMENSION_ID_ORGUNIT } from '@dhis2/analytics'
import i18n from '@dhis2/d2-i18n'
import {
    Button,
    ButtonStrip,
    Modal,
    ModalActions,
    ModalContent,
    ModalTitle,
    Radio,
    Transfer,
} from '@dhis2/ui'

import classes from './styles/FilterSettingsDialog.module.css'

const RadioOptions = ({
    filtersSelectable,
    updateFilterDimensionRestrictability,
}) => (
    <form>
        <div className={classes.radioButton}>
            <Radio
                dense
                checked={!filtersSelectable}
                label={i18n.t('Allow filtering by all dimensions')}
                name="radio-allow-filtering-on-all"
                onChange={payload => {
                    return updateFilterDimensionRestrictability(
                        payload.value === 'true'
                    )
                }}
                value={'false'}
            />
        </div>
        <div className={classes.radioButton}>
            <Radio
                dense
                checked={filtersSelectable}
                label={i18n.t('Only allow filtering by selected dimensions')}
                name="radio-restrict-filtering"
                onChange={payload => {
                    return updateFilterDimensionRestrictability(
                        payload.value === 'true'
                    )
                }}
                value={'true'}
            />
        </div>
    </form>
)

RadioOptions.propTypes = {
    filtersSelectable: PropTypes.bool,
    updateFilterDimensionRestrictability: PropTypes.func,
}

const FilterSettingsDialog = ({
    dimensions,
    restrictFilters,
    initiallySelectedItems,
    onClose,
    onConfirm,
    open,
}) => {
    const [selected, setSelected] = useState(initiallySelectedItems)
    const [filtersSelectable, setFiltersSelectable] = useState(restrictFilters)

    const updateFilterDimensionRestrictability = val => {
        if (val) {
            const otherItems = selected.filter(
                i => i !== DIMENSION_ID_ORGUNIT && i !== DIMENSION_ID_PERIOD
            )
            updateSelectedFilters([
                DIMENSION_ID_PERIOD,
                DIMENSION_ID_ORGUNIT,
                ...otherItems,
            ])
        }
        setFiltersSelectable(val)
    }

    const updateSelectedFilters = sItems => {
        // ensure that orgUnit, period appear before other items
        const periodItem = [
            ...new Set(sItems.filter(i => i === DIMENSION_ID_PERIOD)),
        ]
        const orgUnitItem = [
            ...new Set(sItems.filter(i => i === DIMENSION_ID_ORGUNIT)),
        ]
        const otherItems = sItems.filter(
            i => i !== DIMENSION_ID_ORGUNIT && i !== DIMENSION_ID_PERIOD
        )
        return setSelected([...periodItem, ...orgUnitItem, ...otherItems])
    }

    const closeDialog = () => {
        setSelected(initiallySelectedItems)
        setFiltersSelectable(restrictFilters)
        onClose()
    }

    return (
        <>
            {open && (
                <Modal large position="top" onClose={closeDialog}>
                    <ModalTitle>
                        {i18n.t('Dashboard filter settings')}
                    </ModalTitle>
                    <ModalContent>
                        <div className={classes.descContainer}>
                            <span>
                                {i18n.t(`Dashboards can be filtered by dimensions to change
                                         the data shown. By default, all dimensions are available
                                         as filters. Alternatively, only selected dimensions can
                                         be made available on a dashboard.`)}
                            </span>
                        </div>
                        <RadioOptions
                            filtersSelectable={filtersSelectable}
                            updateFilterDimensionRestrictability={
                                updateFilterDimensionRestrictability
                            }
                        />
                        {filtersSelectable && (
                            <Transfer
                                filterable
                                height="400px"
                                leftHeader={
                                    <div className={classes.transferHeader}>
                                        <span>
                                            {i18n.t('Available Filters')}
                                        </span>
                                    </div>
                                }
                                maxSelections={Infinity}
                                onChange={payload => {
                                    return updateSelectedFilters(
                                        payload.selected
                                    )
                                }}
                                options={dimensions.map(d => {
                                    return { label: d.name, value: d.id }
                                })}
                                optionsWidth="350px"
                                rightHeader={
                                    <div className={classes.transferHeader}>
                                        <span>
                                            {i18n.t('Selected Filters')}
                                        </span>
                                    </div>
                                }
                                selected={selected}
                                selectedWidth="350px"
                            />
                        )}
                    </ModalContent>
                    <ModalActions>
                        <ButtonStrip end>
                            <Button
                                onClick={closeDialog}
                                secondary
                                type="button"
                            >
                                {i18n.t('Cancel')}
                            </Button>
                            <Button
                                onClick={() => {
                                    if (!filtersSelectable) {
                                        setSelected([])
                                    }
                                    onConfirm(filtersSelectable, selected)
                                }}
                                primary
                                type="button"
                            >
                                {i18n.t('Confirm')}
                            </Button>
                        </ButtonStrip>
                    </ModalActions>
                </Modal>
            )}
        </>
    )
}

FilterSettingsDialog.propTypes = {
    dimensions: PropTypes.array,
    initiallySelectedItems: PropTypes.array,
    open: PropTypes.bool,
    restrictFilters: PropTypes.bool,
    onClose: PropTypes.func,
    onConfirm: PropTypes.func,
}

export default FilterSettingsDialog
