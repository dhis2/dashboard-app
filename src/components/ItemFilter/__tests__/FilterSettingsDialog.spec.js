import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import FilterSettingsDialog from '../FilterSettingsDialog'

jest.mock('../../../modules/useDimensions', () => ({
    __esModule: true,
    default: () => {
        return [
            { id: 'kl', name: 'kvikk lunsj' },
            { id: 'ss', name: 'salt sild' },
            { id: 'sm', name: 'seigmenn' },
        ]
    },
}))

const mockSelectedItems = ['kl']

it('renders correctly when filters are not restricted', () => {
    const container = shallow(
        <FilterSettingsDialog
            restrictFilters={false}
            initiallySelectedItems={[]}
            open={true}
        />
    )
    expect(toJson(container)).toMatchSnapshot()
})

it('renders correctly when filters are restricted', () => {
    const container = shallow(
        <FilterSettingsDialog
            restrictFilters={true}
            initiallySelectedItems={mockSelectedItems}
            open={true}
        />
    )
    expect(toJson(container)).toMatchSnapshot()
})
