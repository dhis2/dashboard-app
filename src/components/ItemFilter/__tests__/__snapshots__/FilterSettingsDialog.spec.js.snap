// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when filters are not restricted 1`] = `
<Fragment>
  <Modal
    dataTest="dhis2-uicore-modal"
    large={true}
    onClose={[Function]}
    position="top"
  >
    <ModalTitle
      dataTest="dhis2-uicore-modaltitle"
    >
      Dashboard filter settings
    </ModalTitle>
    <ModalContent
      dataTest="dhis2-uicore-modalcontent"
    >
      <div
        className="descContainer"
      >
        <span>
          Dashboards can be filtered by dimensions to change
                                         the data shown. By default, all dimensions are available
                                         as filters. Alternatively, only selected dimensions can
                                         be made available on a dashboard.
        </span>
      </div>
      <RadioOptions
        filtersSelectable={false}
        updateFilterDimensionRestrictability={[Function]}
      />
    </ModalContent>
    <ModalActions
      dataTest="dhis2-uicore-modalactions"
    >
      <ButtonStrip
        dataTest="dhis2-uicore-buttonstrip"
        end={true}
      >
        <Button
          dataTest="dhis2-uicore-button"
          onClick={[Function]}
          secondary={true}
          type="button"
        >
          Cancel
        </Button>
        <Button
          dataTest="dhis2-uicore-button"
          onClick={[Function]}
          primary={true}
          type="button"
        >
          Confirm
        </Button>
      </ButtonStrip>
    </ModalActions>
  </Modal>
</Fragment>
`;

exports[`renders correctly when filters are restricted 1`] = `
<Fragment>
  <Modal
    dataTest="dhis2-uicore-modal"
    large={true}
    onClose={[Function]}
    position="top"
  >
    <ModalTitle
      dataTest="dhis2-uicore-modaltitle"
    >
      Dashboard filter settings
    </ModalTitle>
    <ModalContent
      dataTest="dhis2-uicore-modalcontent"
    >
      <div
        className="descContainer"
      >
        <span>
          Dashboards can be filtered by dimensions to change
                                         the data shown. By default, all dimensions are available
                                         as filters. Alternatively, only selected dimensions can
                                         be made available on a dashboard.
        </span>
      </div>
      <RadioOptions
        filtersSelectable={true}
        updateFilterDimensionRestrictability={[Function]}
      />
      <Transfer
        dataTest="dhis2-uicore-transfer"
        filterCallback={[Function]}
        filterCallbackPicked={[Function]}
        filterable={true}
        height="400px"
        initialSearchTerm=""
        initialSearchTermPicked=""
        leftHeader={
          <div
            className="transferHeader"
          >
            <span>
              Available Filters
            </span>
          </div>
        }
        maxSelections={Infinity}
        onChange={[Function]}
        options={
          Array [
            Object {
              "label": "kvikk lunsj",
              "value": "kl",
            },
            Object {
              "label": "salt sild",
              "value": "ss",
            },
            Object {
              "label": "seigmenn",
              "value": "sm",
            },
          ]
        }
        optionsWidth="350px"
        renderOption={[Function]}
        rightHeader={
          <div
            className="transferHeader"
          >
            <span>
              Selected Filters
            </span>
          </div>
        }
        selected={
          Array [
            "kl",
          ]
        }
        selectedWidth="350px"
      />
    </ModalContent>
    <ModalActions
      dataTest="dhis2-uicore-modalactions"
    >
      <ButtonStrip
        dataTest="dhis2-uicore-buttonstrip"
        end={true}
      >
        <Button
          dataTest="dhis2-uicore-button"
          onClick={[Function]}
          secondary={true}
          type="button"
        >
          Cancel
        </Button>
        <Button
          dataTest="dhis2-uicore-button"
          onClick={[Function]}
          primary={true}
          type="button"
        >
          Confirm
        </Button>
      </ButtonStrip>
    </ModalActions>
  </Modal>
</Fragment>
`;
